File: quoble/src/main/java/Q1.java
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;

public class Q1 {


    public static void main(String... args){
        SparkSession spark = SparkSession
                .builder()
                .appName("Java Spark SQL basic example")
                .getOrCreate();
        //configure the aws client
        spark.sparkContext().hadoopConfiguration().set("fs.s3a.access.key", "<YOUR_ACCESS_KEY>"); //fill in real values
        spark.sparkContext().hadoopConfiguration().set("fs.s3a.secret.key", "<YOUR_SECRET_KEY>");
        spark.sparkContext().hadoopConfiguration().set("fs.s3a.endpoint", "s3.eu-central-1.amazonaws.com");

        runQ6(spark);

        spark.stop();
    }
    private static void runQ6(SparkSession spark){
        String path="s3://smile-tcp/Input/lineitem.tbl.*";
        //create schema for the Dataset
        StructType schema=DataTypes.createStructType(new StructField[] {
                DataTypes.createStructField("L_ORDERKEY",  DataTypes.IntegerType, true),
                DataTypes.createStructField("L_PARTKEY", DataTypes.IntegerType, true),
                DataTypes.createStructField("L_SUPPKEY", DataTypes.IntegerType, true),
                DataTypes.createStructField("L_LINENUMBER", DataTypes.IntegerType, true),
                DataTypes.createStructField("L_QUANTITY", DataTypes.DoubleType, true),
                DataTypes.createStructField("L_EXTENDEDPRICE", DataTypes.DoubleType, true),
                DataTypes.createStructField("L_DISCOUNT", DataTypes.DoubleType, true),
                DataTypes.createStructField("L_TAX", DataTypes.DoubleType, true),
                DataTypes.createStructField("L_RETURNFLAG", DataTypes.StringType, true),
                DataTypes.createStructField("L_LINESTATUS", DataTypes.StringType, true),
                DataTypes.createStructField("L_SHIPDATE", DataTypes.DateType, true),
                DataTypes.createStructField("L_COMMITDATE", DataTypes.DateType, true),
                DataTypes.createStructField("L_RECEIPTDATE", DataTypes.DateType, true),
                DataTypes.createStructField("L_SHIPINSTRUCT", DataTypes.StringType, true),
                DataTypes.createStructField("L_SHIPMODE", DataTypes.StringType, true),
                DataTypes.createStructField("L_COMMENT", DataTypes.StringType, true),

        });

        //load intput data to the dataset
        Dataset<Row> dataset = spark.read().format("csv").option("delimiter", "|").schema(schema).load(path);
        //create a view for spark sql to work on
        dataset.createOrReplaceTempView("lineitem");
        //sql query Q6
        String query = "select l_returnflag, l_linestatus, sum(l_quantity) as sum_qty, sum(l_extendedprice) as sum_base_price, sum(l_extendedprice*(1-l_discount)) as sum_disc_price, sum(l_extendedprice*(1-l_discount)*(1+l_tax)) as sum_charge,avg(l_quantity) as avg_qty, avg(l_extendedprice) as avg_price,avg(l_discount) as avg_disc, count(*) as count_order from lineitem where l_shipdate <= date '1998-09-01' group by l_returnflag, l_linestatus order by l_returnflag, l_linestatus";
        //run the query
        Dataset<Row> result= spark.sql(query);
        //show results
        result.show();

    }
}

File: pywren/q1.py
import pywren
import csv
from datetime import date,timedelta
import boto3


# get a handle on s3
s3 = boto3.resource(u's3')

# get a handle on the bucket that holds your file
bucket = s3.Bucket(u'smile-tcp')

# get a handle on the object in the Input Folder
prefix_objs = bucket.objects.filter(Prefix="Input")

# get the keys of the files we want to read
keys=[]
for obj in prefix_objs:
    keys.append(obj.key)

#get a pywren executor
wrenexec = pywren.default_executor()

def q1(key):
    #new s3 resource to work on
    s3 = boto3.resource(u's3')
    #read the file from s3 and split it on newline 
    lines=s3.Object("smile-tcp", key).get()['Body'].read().decode('utf-8').split("\n")
    print('all lines of file {} read'.format(key))

    #read the lines into a dicct format to work on
    rows= csv.DictReader(lines,delimiter='|', fieldnames=['L_ORDERKEY', 'L_PARTKEY', 'L_SUPPKEY', 'L_LINENUMBER', 'L_QUANTITY', 'L_EXTENDEDPRICE', 'L_DISCOUNT', 'L_TAX', 'L_RETURNFLAG', 'L_LINESTATUS', 'L_SHIPDATE', 'L_COMMITDATE', 'L_RECEIPTDATE', 'L_SHIPINSTRUCT', 'L_SHIPMODE', 'L_COMMENT'])
    #print('all rows of file {} read'.format(key))
    #run the query for every line in the file
    q_date = date(1998,12,1) - timedelta(days=80)
    grouped={}
    for row in rows:
        parsed_date=row['L_SHIPDATE'].split("-")
        shipdate= date(int(parsed_date[0]),int(parsed_date[1]),int(parsed_date[2]))
        
        if  shipdate <= q_date:
            key = row['L_RETURNFLAG']+row['L_LINESTATUS']
            grouped.setdefault(key,[]).append([float(row['L_QUANTITY']), float(row['L_EXTENDEDPRICE']),float(row['L_DISCOUNT']),float(row['L_TAX'])])
    #filter None results for sum function
    
    #group the results by returnflag and linestatus
        
    results=[]
    #do all the math in the query
    for key in grouped:
        sum_qty=0
        sum_base_price=0
        sum_disc_price=0
        sum_charge=0
        sum_disc=0
        count=0

        for row in grouped[key]:
            sum_qty+=row[0]
            sum_base_price+=row[1]
            sum_disc_price +=row[1]*(1-row[2])
            sum_charge+=row[1]*(1-row[2])*(1+row[3])
            sum_disc+=row[2]
            count+=1
        results.append([key,sum_qty,sum_base_price,sum_disc_price,sum_charge,sum_disc,count])
    return results
#reduce the grouped results per file
def q1_reduce(raw):

    grouped={}
    for part in raw:
        for group in part:
            grouped.setdefault(group[0],[]).append(group[1:])
    results=[]
    for key in grouped:
        sum_qty=0
        sum_base_price=0
        sum_disc_price=0
        sum_charge=0
        sum_disc=0
        count=0
        for row in grouped[key]:
            sum_qty+=row[1]
            sum_base_price+=row[2]
            sum_disc_price +=row[3]
            sum_charge+=row[4]
            sum_disc+=row[5]
            count+=row[-1]
        results.append([key,sum_qty,sum_base_price,sum_disc_price,sum_charge,sum_qty/count,sum_base_price/count,sum_disc/count,count])
    return results
#run the query on pywren keys[0] is the Input folder so we ignore this one 
futures = wrenexec.map(q1, keys[1:])

#wait for futures to fill
raw=pywren.get_all_results(futures)
#reduce the results
res=q1_reduce(raw)
#format the output to fit the query
for key in sorted(res):
    print('group:{}, sum_qty:{:.2f}, sum_base_price:{:.2f}, sum_disc_price:{:.2f}, sum_charge:{:.2f}, avg_qty:{:.2f}, avg_price:{:.2f}, avg_disc:{:.2f}, count_order:{}'.format(key[0],key[1],key[2],key[3],key[4],key[5],key[6],key[7],key[8]))